### PR TITLE
feat(remap): Add `parse_query_string` function

### LINKED ISSUE #5772: New `parse_query_string` Remap function
Parses a URL encoded query string.

## Example

```json
{
	"query": "foo=1&bar=2"
}
```

And remap:

```js
.parsed = parse_query_string(.query)
```

Would result in:

```json
{
	"query": "foo=1&bar=2",
	"parsed": {
		"foo": "1",
		"bar": "2"
	}
}
```

### PR DESCRIPTION
**Pull request by vladimir-dd**
Closes #5772 

Signed-off-by: Vladimir Zhuk <vladimir.zhuk@datadoghq.com>

### CODE REVIEW

**Review by jszwedko — COMMENTED** [review]
Thanks @vladimir-dd !

Do you mind adding a benchmark here as well: https://github.com/timberio/vector/blob/234e4665ad0c80fee1c4dfeafe1887e97a012c7a/lib/vrl/stdlib/benches/benches.rs

File: lib/vrl/stdlib/src/parse_query_string.rs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            keyword: "target",
+            kind: kind::ANY,
+            required: true,
+        }]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseQueryStringFn {
+    target: expression::Query,
```
Comment (jszwedko) [inline lib/vrl/stdlib/src/parse_query_string.rs:41]: We talked on Slack that this should be a string value rather than requiring a object path to be specified. Once that change is made, we may want to rename `target` -> `value` to match other `parse_X` functions.

Comment (vladimir-dd) [reply lib/vrl/stdlib/src/parse_query_string.rs]: Yep, misunderstood the requirements first, updated. Thanks!

**Review by ktff — APPROVED** [review]
LGTM

**Review by jszwedko — APPROVED** [review]
Looks great, thanks @vladimir-dd !

File: lib/vrl/stdlib/src/parse_query_string.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+        Ok(Box::new(ParseQueryStringFn { value }))
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            kind: kind::ANY,
```
Comment (FungusHumungus) [inline lib/vrl/stdlib/src/parse_query_string.rs:34]: The parameter kind should be `kind::BYTES`. It should only accept Strings.

File: lib/vrl/stdlib/src/parse_query_string.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+struct ParseQueryStringFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for ParseQueryStringFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let query_string = self.value.resolve(ctx)?.try_bytes()?;
+        let result = form_urlencoded::parse(query_string.as_ref())
+            .map(|(k, v)| (k.to_string(), v.into()))
+            .collect();
```
Comment (FungusHumungus) [inline lib/vrl/stdlib/src/parse_query_string.rs:50]: Our other parse functions will return an error if it can't actually parse any data. I think we should do the same here.

What does `form_urlencoded::parse` return if we pass a string that isn't `key=value`, eg `form_urlencoded::parse("thing")`? Does it ignore that data? Does it become the key with an empty value?

If it does ignore that data, then I think we should raise an error here, so if `result.len() == 0` for any parameter other than an empty string, we should error.

File: lib/vrl/stdlib/src/parse_query_string.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        Ok(Value::Object(result))
+    }
+
+    fn type_def(&self, _: &state::Compiler) -> TypeDef {
+        TypeDef::new().fallible().object::<(), Kind>(type_def())
```
Comment (FungusHumungus) [inline lib/vrl/stdlib/src/parse_query_string.rs:55]: If you find we don't need to raise an error re.. the above comment, then this function should be `infallible()`.

File: lib/vrl/stdlib/src/parse_query_string.rs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    test_function![
+        parse_query_string => ParseQueryString;
+
+        type_def {
+            args: func_args![value: value!("foo=%2B1&bar=2")],
+            want: Ok(value!({
+                foo: "+1",
+                bar: "2",
+            })),
+            tdef: TypeDef::new().fallible().object::<(), Kind>(type_def()),
```
Comment (FungusHumungus) [inline lib/vrl/stdlib/src/parse_query_string.rs:78]: Can you add a test here to test what happens if the parameter is just 

```
value!("something")
```

**Review by StephenWakely — CHANGES_REQUESTED** [review]
Beautiful! Some tiny changes and a question around the error path and we should be good to go!

File: lib/vrl/stdlib/src/parse_query_string.rs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use shared::btreemap;
+
+    #[test]
+    fn parse_query_string() {
+        let cases = vec![(
```
[the remark this answers is not in the document]
Comment (FungusHumungus) [reply lib/vrl/stdlib/src/parse_query_string.rs]: The `test_type_def` tests are like totally February 2021.. These tests have been moved into the `tdef` field of `test_function!` now.

File: lib/vrl/stdlib/src/parse_query_string.rs (hunk 3 above)
Comment (vladimir-dd) [reply lib/vrl/stdlib/src/parse_query_string.rs:50]: Good point! It returns a key with an empty value, so I added more tests and made this function `infallible`. Also updated the doc.

File: lib/vrl/stdlib/src/parse_query_string.rs (hunk 5 above)
Comment (vladimir-dd) [reply lib/vrl/stdlib/src/parse_query_string.rs]: Done.

**Review by StephenWakely — APPROVED** [review]
Nice!

File: lib/vrl/stdlib/src/parse_query_string.rs (hunk 3 above)
Comment (jszwedko) [reply lib/vrl/stdlib/src/parse_query_string.rs:50]: Just circling back to this. To @FungusHumungus's point, we've typically had the `parse_` functions (like `parse_grok`) fail if the provided input doesn't match the pattern. I could see doing the same here and returning an error if given an input that is not a valid query string.

Comment (vladimir-dd) [reply lib/vrl/stdlib/src/parse_query_string.rs]: @jszwedko I added more tests(and updated the doc) to show that we support empty keys and values(I believe empty values must be allowed and empty keys can be filtered by users with `compact`). Given that, I'm struggling to think of any invalid input, therefore I made this function `infallible`. Do you see any example of invalid input, which would require error handling?

Comment (jszwedko) [reply lib/vrl/stdlib/src/parse_query_string.rs]: 👍 I didn't realize empty values were allowed. That makes sense.

**Review by jszwedko — APPROVED** [review]

File: docs/reference/remap/functions/parse_query_string.cue [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+		},
+	]
+	internal_failure_reasons: []
+	return: types: ["object"]
+
+	examples: [
+		{
+			title: "Parse query string"
+			source: #"""
+				parse_query_string("foo=%2B1&bar=2&bar=3&xyz")
```
Comment (JeanMertz) [inline docs/reference/remap/functions/parse_query_string.cue:29]: Ruby on Rails (😱 @binarylogic) uses (used?) query strings like this for arrays: `?ids[]=1025&ids[]=1027`. Should we special-case that to group them into an array of `ids` similar to how we group `bar` in this example?

Alternatively, we keep things as they are, but add a (tested) example to the documentation that shows how this works.

Comment (JeanMertz) [inline docs/reference/remap/functions/parse_query_string.cue:29]: Also, should we strip `?` as a leading character for convenience? I can see that being a case that people might run into, and I don't see a particular downside to always doing so.

File: lib/vrl/stdlib/src/parse_query_string.rs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+#[derive(Debug, Clone)]
+struct ParseQueryStringFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for ParseQueryStringFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let query_string = self.value.resolve(ctx)?.try_bytes()?;
+        let mut result: BTreeMap<String, Value> = BTreeMap::new();
```
Comment (JeanMertz) [inline lib/vrl/stdlib/src/parse_query_string.rs:50]:
```suggestion
        let mut result = BTreeMap::new();
```

The compiler _should_ be able to derive that this type is `BTreeMap<String, Value>`, so this type definition can be removed.

File: lib/vrl/stdlib/src/parse_query_string.rs [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        let parsed = form_urlencoded::parse(query_string.bytes());
```
Comment (JeanMertz) [inline lib/vrl/stdlib/src/parse_query_string.rs:51]:
```suggestion
        let parsed = form_urlencoded::parse(&query_string);
```

Dereferencing `Byte` type results in `&[u8]`, which should work here.

Also, this function got [removed](https://docs.rs/bytes/1.0.1/bytes/struct.Bytes.html#method.bytes) in the 1.x release (we're still on 0.6 until Tokio is upgraded), so it might be good to just do that straight away.

File: lib/vrl/stdlib/src/parse_query_string.rs [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        for (k, v) in parsed {
+            let value = v.as_ref().into();
```
Comment (JeanMertz) [inline lib/vrl/stdlib/src/parse_query_string.rs:53]:
```suggestion
            let value = v.into();
```

`vrl::Value` implements `From<Cow<'_, str>>`, so I suspect `v.into()` would work as well here.

File: lib/vrl/stdlib/src/parse_query_string.rs [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+            let entry = result.entry(k.as_ref().to_owned());
```
Comment (JeanMertz) [inline lib/vrl/stdlib/src/parse_query_string.rs:54]:
```suggestion
            let entry = result.entry(k.into_owned());
```

`Cow` has a [`into_owned`](http://doc.rust-lang.org/1.50.0/alloc/borrow/enum.Cow.html#method.into_owned), which does the same.

File: lib/vrl/stdlib/src/parse_query_string.rs [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+                        e.get_mut().as_array_mut().unwrap().push(value);
+                    } else {
+                        let prev_value = e.get().to_owned();
+                        result.insert(k.as_ref().into(), vec![prev_value, value].into());
+                    }
+                }
+                Entry::Vacant(e) => {
+                    e.insert(value);
+                }
+            }
```
Comment (JeanMertz) [inline lib/vrl/stdlib/src/parse_query_string.rs:67]:
```suggestion
            result
                .entry(k.into_owned())
                .and_modify(|v| match v {
                    Value::Array(v) => v.push(value),
                    v => value!([v, value]),
                })
                .or_insert(value);
```

_Something_ like this should work, and seems a bit nicer to read?

File: lib/vrl/stdlib/src/parse_query_string.rs [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+        }
+        Ok(Value::Object(result))
```
Comment (JeanMertz) [inline lib/vrl/stdlib/src/parse_query_string.rs:69]: nit

```suggestion
        Ok(result.into())
```

**Review by JeanMertz — APPROVED** [review]
Nice!

Left a few nits, but other than that, this looks good to me.

File: docs/reference/remap/functions/parse_query_string.cue (hunk 7 above)
Comment (vladimir-dd) [reply docs/reference/remap/functions/parse_query_string.cue:29]: Good idea! Done.

File: lib/vrl/stdlib/src/parse_query_string.rs (hunk 9 above)
Comment (vladimir-dd) [reply lib/vrl/stdlib/src/parse_query_string.rs]: Thanks, corrected.

File: lib/vrl/stdlib/src/parse_query_string.rs (hunk 8 above)
Comment (vladimir-dd) [reply lib/vrl/stdlib/src/parse_query_string.rs:55]: It was complaining here:
```
|         let mut result = BTreeMap::new();
   |             ---------- consider giving `result` the explicit type `BTreeMap<std::string::String, V>`, where the type parameter `V` is specified
...
62 |                     if e.get().is_array() {
   |                          ^^^ cannot infer type for type parameter `V`
```

File: lib/vrl/stdlib/src/parse_query_string.rs (hunk 10 above)
Comment (vladimir-dd) [reply lib/vrl/stdlib/src/parse_query_string.rs:58]: Indeed 👍

File: lib/vrl/stdlib/src/parse_query_string.rs (hunk 11 above)
Comment (vladimir-dd) [reply lib/vrl/stdlib/src/parse_query_string.rs:59]: Thanks, updated.

File: lib/vrl/stdlib/src/parse_query_string.rs (hunk 13 above)
Comment (vladimir-dd) [reply lib/vrl/stdlib/src/parse_query_string.rs:74]: Thx, updated.

File: lib/vrl/stdlib/src/parse_query_string.rs (hunk 12 above)
Comment (vladimir-dd) [reply lib/vrl/stdlib/src/parse_query_string.rs]: Much nicer, I remember I was struggling to make it work first and gave up thinking it is impossible, now I hope I got it right - @JeanMertz could you take a look again please to make sure it follows rust conventions/is optimal?

File: lib/vrl/stdlib/src/parse_query_string.rs (hunk 8 above)
Comment (vladimir-dd) [reply lib/vrl/stdlib/src/parse_query_string.rs]: Works now after refactoring.

File: docs/reference/remap/functions/parse_query_string.cue (hunk 7 above)
Comment (vladimir-dd) [reply docs/reference/remap/functions/parse_query_string.cue:29]: Added an example/test - we'd parse it as `"ids[]" = ["1025", "1027"]`. Since it's a RoR-specific way of handling array parameters I would keep it simple. WDYT?

**Review by JeanMertz — APPROVED** [review]
Sweet! LGTM
